### DISCUSSION TITLE
Wix v4 - Error WIX0044	The CustomAction element's BinaryRef, Directory, Error, FileRef, Property, or Script attribute was not found

### DISCUSSION THREAD

**Discussion by garevaul** [wixtoolset/issues#7346] [Questions] [answered]
I upgraded to Wix4, but I am getting this Error.

I had this code working without the CustomAction in Wix3

But I have added the Custom Action now, since I was getting the Error for Publish Event.

But, what is the Correct way for Wix4 ?

And I have 2 Statements (One Commented) for 'Publish Event'. Which is the Correct one ? I had the Uncommented one working fine earlier.

```
<Control Id="Print" Type="PushButton" X="88" Y="243" Width="56" Height="17" Text="!(loc.WixUIPrint)">
   <Publish Event="DoAction" Value="WixUIPrintEula_$(sys.BUILDARCHSHORT)_$(sys.BUILDARCHSHORT)" />
   <!--<Publish Event="DoAction" Value="WixUIPrintEula">1</Publish>-->
</Control>

```
`<CustomAction Id="WixUIPrintEula" BinaryKey="WixUIWixca" DllEntry="PrintEula" Return="ignore" Execute="immediate" />
`			

Error	WIX0004	The CustomAction element contains an unexpected attribute 'BinaryKey'.
Error	WIX0044	The CustomAction element's BinaryRef, Directory, Error, FileRef, Property, or Script attribute was not found; one of these is required.

Regards

**Comment by barnson** [comment, marked as answer]
https://wixtoolset.org/docs/fourthree/faqs/#converting-custom-wixui-dialog-sets

**Reply by garevaul** [reply]
@barnson thanx for the link.

I did this change and it compiles fine. After I Click on the Print, it Prints also fine.

```
<Control Id="Print" Type="PushButton" X="88" Y="243" Width="56" Height="17" Text="!(loc.WixUIPrint)">
    <?foreach WIXUIARCH in X86;X64;A64 ?>
		<Publish Event="DoAction" Value="WixUIPrintEula_$(WIXUIARCH)"/>
	<?endforeach?>
</Control>
```

BUT it Shows me this Windows Print Dialog 2 times. So I have to Click On Print the first time, and then I have to Click on Cancel. And Instead of Clicking on Cancel the 2nd time, if I click on Print again, then also the Windows Print Dialog Goes away and takes me back to the Wix Installer Dialog.

![image](https://user-images.githubusercontent.com/15125247/229334911-48d19324-a36e-4b51-868a-b6bd7f9c993c.png)


Regards

**Reply by barnson** [reply]
Here's a working example you can compare to: https://github.com/wixtoolset/wix4/tree/develop/src/ext/UI/test/WixToolsetTest.UI/TestData/InstallDir_SpecialDlg

**Comment by garevaul** [comment]
@barnson Thanx for the Link. I made the relevant changes and It worked.

I was actually doing the '' within the UI>Dialog>Control Element, whereas it should actually be even outside the UI element, as a separate Entity, as a Separate Fragment.

Now the Windows Print Dialog just once.

Regards
@garevaul
